controller/engine_controller.go

```diff
@@ -987,21 +987,23 @@ func (m *EngineMonitor) refresh(engine *longhorn.Engine) error {
 				}
 			}
 
-			engine.Status.SnapshotMaxCount = volumeInfo.SnapshotMaxCount
-			if engine.Spec.SnapshotMaxCount != volumeInfo.SnapshotMaxCount {
-				logrus.Infof("Correcting flag SnapshotMaxCount from %d to %d", volumeInfo.SnapshotMaxCount, engine.Spec.SnapshotMaxCount)
-				if err := engineClientProxy.VolumeSnapshotMaxCountSet(engine); err != nil {
-					return errors.Wrapf(err, "failed to correct flag SnapshotMaxCount from %d to %d",
-						volumeInfo.SnapshotMaxCount, engine.Spec.SnapshotMaxCount)
+			if cliAPIVersion >= 10 && im.Status.APIVersion >= 5 {
```

Comment:
Is it necessary to nest inside `if cliAPIVersion >= 7 && im.Status.APIVersion >= 3 {`?

Suggest adding a brief description for clarity.

Reply:
I'm not sure in which situation we want to set `CurrentSize` as `VolumeSize` and remove `RebuildStatus`. If it's a case for `cliAPIVersion <= 6 || im.Status.APIVersion <= 2`, I think we may need nest here.

Reply:
Consider doing this. They are both still in the if statement that you've mentioned.
```go
		if cliAPIVersion >= 7 && im.Status.APIVersion >= 3 {
			// do something
		}

		if cliAPIVersion >= 10 && im.Status.APIVersion >= 5 {
			// do something
		}
```

Reply:
In this case, we can't rely on `else` condition. We have to change it like this:

```go
        if cliAPIVersion >= 7 && im.Status.APIVersion >= 3 {
                // do something
        }

        if cliAPIVersion >= 9 && im.Status.APIVersion >= 5 {
                // do something
        }

        if cliAPIVersion <=6 || im.Status.APIVersion <= 2 {
                // For incompatible running engine, the current size is always `engine.Spec.VolumeSize`. 
 	        engine.Status.CurrentSize = engine.Spec.VolumeSize 
 	        engine.Status.RebuildStatus = map[string]*longhorn.RebuildStatus{}
        }
```

Reply:
Let's tackle this at https://github.com/longhorn/longhorn/issues/7557.

Reply:
> In this case, we can't rely on `else` condition. We have to change it like this:
> 
> ```go
>         if cliAPIVersion >= 7 && im.Status.APIVersion >= 3 {
>                 // do something
>         }
> 
>         if cliAPIVersion >= 9 && im.Status.APIVersion >= 5 {
>                 // do something
>         }
> 
>         if cliAPIVersion <=6 || im.Status.APIVersion <= 2 {
>                 // For incompatible running engine, the current size is always `engine.Spec.VolumeSize`. 
>  	        engine.Status.CurrentSize = engine.Spec.VolumeSize 
>  	        engine.Status.RebuildStatus = map[string]*longhorn.RebuildStatus{}
>         }
> ```

I am thinking it will be more like:
```go
	if (datastore.IsDataEngineV1(engine.Spec.DataEngine) && cliAPIVersion >= emeta.CLIAPIMinVersion) ||
		datastore.IsDataEngineV2(engine.Spec.DataEngine) {

		if cliAPIVersion >= 7 && im.Status.APIVersion >= 3 {
			// do something
		}

		if cliAPIVersion >= 9 && im.Status.APIVersion >= 5 {
			// do something
		}
	} else {
		engine.Status.CurrentSize = engine.Spec.VolumeSize
		engine.Status.RebuildStatus = map[string]*longhorn.RebuildStatus{}
	}
```
Comparing to:
```go
	if (datastore.IsDataEngineV1(engine.Spec.DataEngine) && cliAPIVersion >= emeta.CLIAPIMinVersion) ||
		datastore.IsDataEngineV2(engine.Spec.DataEngine) {

		if cliAPIVersion >= 7 && im.Status.APIVersion >= 3 {

			// do something

			if cliAPIVersion >= 9 && im.Status.APIVersion >= 5 {
				// do something
			}
		}
	} else {
		engine.Status.CurrentSize = engine.Spec.VolumeSize
		engine.Status.RebuildStatus = map[string]*longhorn.RebuildStatus{}
	}
```

Reply:
Thank you. I misread the `else` block. I thought it was with  `cliAPIVersion >= 7 && im.Status.APIVersion >= 3` condition 😅. Yeah, it's good to move `cliAPIVersion >= 9 && im.Status.APIVersion >= 5` out.